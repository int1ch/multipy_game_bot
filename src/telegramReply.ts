import {
  ParseMode,
  ReplyKeyboardMarkup,
  ReplyKeyboardRemove,
} from "@grammyjs/types";
import { AnyAaaaRecord } from "dns";

import { Context, Keyboard } from "grammy";
import { GameResponse } from "./GameProcessor";

type ReplyMethodParams = Parameters<Context["reply"]>;
type ReplyReturnType = ReturnType<Context["reply"]>;
type ReplyMethod = (...args: ReplyMethodParams) => ReplyReturnType;

interface ContextWithReply {
  reply: ReplyMethod;
}

//output
export async function telegramReply(
  ctx: ContextWithReply,
  response: GameResponse
) {
  //const ctx = this.ctx;
  const text = response.text;

  const otherOptions: Parameters<typeof ctx["reply"]>[1] = {};
  if (response.variants) {
    otherOptions.reply_markup = {
      one_time_keyboard: true,
      keyboard: telegramReplyKeyboard(response.variants),
    };
  }
  if (response.parse_mode) {
    otherOptions.parse_mode = response.parse_mode;
  }
  if (response.removeKeyboard) {
    otherOptions.reply_markup = { remove_keyboard: true };
  }
  return await ctx.reply(text, otherOptions);
}

export function telegramReplyKeyboard(variants: string[]) {
  const keyboard = new Keyboard();
  for (const variant of variants) {
    keyboard.text(variant);
  }
  return keyboard.build();
}
