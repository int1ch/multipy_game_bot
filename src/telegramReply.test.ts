import { telegramReply } from "./telegramReply";

function getMockedContext() {
  const replyMock = jest.fn();
  replyMock.mockResolvedValue(undefined);
  const context = {
    reply: replyMock,
  };
  return [context, replyMock] as const;
}
test("telegram  text replay", () => {
  const [context, replyMock] = getMockedContext();
  telegramReply(context, { text: "HelloWorld" });
  expect(replyMock).toBeCalledTimes(1);
  expect(replyMock).toBeCalledWith("HelloWorld", {});
});
test("telegram  text replay variants", () => {
  const [context, replyMock] = getMockedContext();
  telegramReply(context, { text: "HelloWorld", variants: ["1", "A"] });
  expect(replyMock).toBeCalledTimes(1);
  expect(replyMock).toMatchSnapshot();
  /*"HelloWorld", {
    reply_markup: {
      keyboard: [
        [
          {
            text: "1",
          },
          {
            text: "A",
          },
        ],
      ],
      one_time_keyboard: true,
    },
  });*/
});

test("telegram  text MarkdownV2", () => {
  const [context, replyMock] = getMockedContext();
  telegramReply(context, { text: "HelloWorld", parse_mode: "MarkdownV2" });
  expect(replyMock).toBeCalledTimes(1);
  expect(replyMock).toBeCalledWith("HelloWorld", {
    parse_mode: "MarkdownV2",
  });
});
