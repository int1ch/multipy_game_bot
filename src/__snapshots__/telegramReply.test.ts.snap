// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`telegram  text replay variants 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "HelloWorld",
      Object {
        "reply_markup": Object {
          "keyboard": Array [
            Array [
              Object {
                "text": "1",
              },
              Object {
                "text": "A",
              },
            ],
          ],
          "one_time_keyboard": true,
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;
